hkmc2/shared/src/test/mlscript/codegen/TraceLog.mls

```diff
@@ -0,0 +1,185 @@
+//│     traceLogEnterMsg = globalThis.String.prototype.concat.call("calling: fib(", traceLogParam_a, ")");
+//│     tmp = globalThis.Predef.TraceLogger.log(traceLogEnterMsg);
+//│     traceLogPrevIndent = globalThis.Predef.TraceLogger.indent();
+//│     scrut = a <= 1;
+//│     if (scrut) {
+//│       tmp1 = a;
+//│     } else {
+//│       tmp2 = a - 1;
+//│       tmp3 = globalThis.fib(tmp2);
+//│       tmp4 = a - 2;
+//│       tmp5 = globalThis.fib(tmp4);
+//│       tmp1 = tmp3 + tmp5;
+//│     }
+//│     traceLogRes = tmp1;
+//│     return traceLogRes;
+//│   } finally {
+//│     traceLogResInspected = globalThis.util.inspect(traceLogRes);
+//│     traceLogRetMsg = globalThis.String.prototype.concat.call("return: ", traceLogResInspected);
+//│     tmp6 = globalThis.Predef.TraceLogger.resetIndent(traceLogPrevIndent);
+//│     tmp7 = globalThis.Predef.TraceLogger.log(traceLogRetMsg);
+//│   }
+//│   
+//│ }
+//│ undefined
+
+fun f(x) = g(x)
+fun g(x) = h(x)
+fun h(x) = fib(x)
+f(5)
+//│ > calling: f(5)
+//│ > | calling: g(5)
+//│ > | | calling: h(5)
+//│ > | | | calling: fib(5)
+//│ > | | | | calling: fib(4)
+//│ > | | | | | calling: fib(3)
+//│ > | | | | | | calling: fib(2)
+//│ > | | | | | | | calling: fib(1)
+//│ > | | | | | | | return: 1
+//│ > | | | | | | | calling: fib(0)
+//│ > | | | | | | | return: 0
+//│ > | | | | | | return: 1
+//│ > | | | | | | calling: fib(1)
+//│ > | | | | | | return: 1
+//│ > | | | | | return: 2
+//│ > | | | | | calling: fib(2)
+//│ > | | | | | | calling: fib(1)
+//│ > | | | | | | return: 1
+//│ > | | | | | | calling: fib(0)
+//│ > | | | | | | return: 0
+//│ > | | | | | return: 1
+//│ > | | | | return: 3
+//│ > | | | | calling: fib(3)
+//│ > | | | | | calling: fib(2)
+//│ > | | | | | | calling: fib(1)
+//│ > | | | | | | return: 1
+//│ > | | | | | | calling: fib(0)
+//│ > | | | | | | return: 0
+//│ > | | | | | return: 1
+//│ > | | | | | calling: fib(1)
+//│ > | | | | | return: 1
+//│ > | | | | return: 2
+//│ > | | | return: 5
+//│ > | | return: 5
+//│ > | return: 5
+//│ > return: 5
+//│ = 5
+
+
+abstract class E: S | N
+class S(x) with
+  fun toString() = "S(" + x + "): " + toInt(succ(x))
+class N() with
+  fun toString() = "N: 0" 
+fun succ(x) = S(x)
+fun toInt(x) = if x is
+  N then 0
+  S(n) then 1 + toInt(n)
+
+
+(S(S(N()))).toString()
+//│ > calling: toString()
+//│ > | calling: toString()
+//│ > | | calling: toString()
+//│ > | | return: 'N: 0'
+//│ > | | calling: succ(N {})
+//│ > | | return: S { x: N {} }
+//│ > | | calling: toInt(S { x: N {} })
+//│ > | | | calling: toInt(N {})
+//│ > | | | return: 0
+//│ > | | return: 1
+//│ > | return: 'S(N: 0): 1'
+//│ > | calling: succ(S { x: N {} })
+//│ > | return: S { x: S { x: N {} } }
+//│ > | calling: toInt(S { x: S { x: N {} } })
+//│ > | | calling: toInt(S { x: N {} })
+//│ > | | | calling: toInt(N {})
+//│ > | | | return: 0
+//│ > | | return: 1
+//│ > | return: 2
+//│ > return: 'S(S(N: 0): 1): 2'
+//│ = 'S(S(N: 0): 1): 2'
+
+
+class Cls(a, b, c, d, e, f)
```

Comment:
I guess the point in having so many fields is to make it break the line. But that makes the output quite noisy, so could you ensure it only prints the thing a couple of times?

Reply:
Thanks! I have pushed the change to modify this test so that only one multiline trace message is shown.